File: c/stack/src/stack.c
#include "stack.h"

int get_size(stack* s) {
    return s->size;
}

int peek(stack* s) {
    return s->head->data;
}

int pop(stack* s) {
    if (s == NULL || s->head == NULL) {
        return -1;
    }
    int ret = s->head->data;
    stack_node* next = s->head->next;
    s->head->data = -1;
    free(s->head);
    s->head = next;
    s->size--;
    next = NULL; // dereference temp pointer
    return ret;
}

int push(stack *s, int n) {
    if (s == NULL) {
        return -1;
    }
    if (s->head == NULL) {
        s->head = malloc(sizeof(stack_node));
        s->head->data = n;
        s->head->next = NULL;
    }
    stack_node* newHead = malloc(sizeof(stack_node));
    newHead->data = n;
    newHead->next = s->head;
    s->head = newHead;
    s->size++;
    return 1;
}

void print_stack(stack* s) {
    char str[1024] = "";
    stack_node* current = s->head;
    printf("%s: ", "Current Stack");
    while (current->next->next != NULL) {
        sprintf(str, "%d->", current->data);
        printf("%s", str);
        current = current->next;
    }
    sprintf(str, "%d", current->data);
    printf("%s\n", str);
}

stack* new_stack(void) {
    stack* s = malloc(sizeof(stack));
    s->size = 0;
    s->head = NULL;
    return s;
}

void delete_stack(stack* s) {
    if (s == NULL) {
        return;
    }
    if (s->head == NULL) {
        free(s);
    }
    while (get_size(s) > 0) {
        pop(s);
    }
    s->head->data = -1;
    free(s->head);
    free(s);
}

File: Stack.java
public class Stack<E> {
    private int size;
    private Node head;

    private class Node {
        Node next;
        E data;

        public Node() {}

        public Node(E data) {
            this.data = data;
        }
    }

    public Stack() {
        this.head = new Node();
    }

    public E peek() { return this.head.data; }

    public int size() { return this.size; }

    public void push(E data) { 
        Node newHead = new Node(data);
        newHead.next = this.head;
        this.head = newHead;
        this.size++;
    }

    public E pop() {
        E removed = this.head.data;
        this.head = this.head.next;
        this.size--;
        return removed;
    }

    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer();
        Node current = head;

        sb.append("[");
        while (current.next != null) { 
            sb.append(current.data);
            sb.append(",");
            current = current.next;
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("]");
        return sb.toString();
    }
 }
